Migrate main readme.md (#59)
App-vNext/Polly-Samples: README.md

![Polly logo](https://raw.github.com/App-vNext/Polly/main/Polly-Logo.png)

> It provides sample implementations using the [Polly library](https://www.github.com/App-vNext/Polly).

Reviewer: ```suggestion
This repository provides sample implementations of using the [Polly library](https://www.github.com/App-vNext/Polly) in a .NET application.
```

The intent of this project is to help newcomers kick-start the use of Polly within their own projects.

The samples demonstrate the policies in action, against faulting endpoints.

> ## Table of contents

Reviewer: I think we can remove this and just lean on the built-in GitHub ToC as then it's one less thing to maintain.

![image](https://github.com/App-vNext/Polly-Samples/assets/1439341/1a597540-fc5d-44ba-ab0c-3cb885a42937)

- [Polly-Samples](#polly-samples)
  - [Table of contents](#table-of-contents)
  - [Projects](#projects)
  - [Demos](#demos)
    - [General information](#general-information)
    - [Sequence](#sequence)
  - [Want further information?](#want-further-information)
  - [Slide decks](#slide-decks)

## Projects

> The solution contains three applications and one class library.

Reviewer: ```suggestion
The solution contains three applications and one class library:

```

> - `PollyTestWebApi`: This application is a web API with three endpoint. ([Further information](/PollyTestWebApi/README.md))

Reviewer: ```suggestion
- `PollyTestWebApi`: This application is a web API with three endpoints. ([Further information](/PollyTestWebApi/README.md))
```

- `PollyDemos`: This library contains the Polly demos. ([Further information](/PollyDemos/README.md))
- `PollyTestClientConsole`: This application provides a CLI to test a demo. ([Further information](/PollyTestClientConsole/README.md))
- `PollyTestClientWpf`: This application provides a GUI to walk through the demos. ([Further information](/PollyTestClientWpf/README.md))

```mermaid
flowchart LR
    console{{PollyTestClientConsole}}
    wpf{{PollyTestClientWPF}}
    lib>PollyDemos]
    api[/PollyTestWebApi\]

    console -- uses --> lib
    wpf -- uses --> lib
    lib -- invokes --> api
```

## Demos

### General information

- The demos run against an example 'faulting server'.
  - To simulate failure, the dummy server rejects more than 3 calls in any five-second period.
- Be sure to read the `<summary>` at the top of each demo.
  - This explains the intent of that demo, and what resilience it adds to its handling of the calls to the 'faulting server'.

> - Sometimes the `<summary>` also highlights what this demo _doesn't_ achieve often picked up in the following demo.

Reviewer: ```suggestion
- Sometimes the `<summary>` also highlights what this demo _doesn't_ achieve, which is often picked up in the following demo.
```

> - Explore the demos in sequence, for best understanding.

Reviewer: ```suggestion
- Explore the demos in sequence for best understanding.
```

| 06 | Wait and retry nesting circuit breaker | [Here](PollyDemos/Sync/Demo06_WaitAndRetryNestingCircuitBreaker.cs) | [Here](PollyDemos/Async/AsyncDemo06_WaitAndRetryNestingCircuitBreaker.cs) |
| 07 | Wait and retry chaining with circuit breaker by using Pipeline | [Here](PollyDemos/Sync/Demo07_WaitAndRetryNestingCircuitBreakerUsingPipeline.cs) | [Here](PollyDemos/Async/AsyncDemo07_WaitAndRetryNestingCircuitBreakerUsingPipeline.cs) |
| 08 | Fallback, Retry, and CircuitBreaker in a Pipeline | [Here](PollyDemos/Sync/Demo08_Pipeline-Fallback-WaitAndRetry-CircuitBreaker.cs) | [Here](PollyDemos/Async/AsyncDemo08_Pipeline-Fallback-WaitAndRetry-CircuitBreaker.cs) |
| 09 | Fallback, Timeout, and Retry in a Pipeline | [Here](PollyDemos/Sync/Demo09_Pipeline-Fallback-Timeout-WaitAndRetry.cs) | [Here](PollyDemos/Async/AsyncDemo09_Pipeline-Fallback-Timeout-WaitAndRetry.cs) |
| 10 | Without isolation: Faulting calls swamp resources, <br/>also prevent good calls | - | [Here](PollyDemos/Async/AsyncDemo10_SharedConcurrencyLimiter.cs) |

> | 11 | With isolation: Faulting calls separated, <br/>do not swamp resources, good calls still succeed | - | [Here](PollyDemos/Async/AsyncDemo11_MultipleConcurrencyLimiters.cs) |

Reviewer: IMHO "Code" would be a better label for the links than "Here".

## Want further information?

> - Any questions about the operation of the demos, ask on this repo.

Reviewer: ```suggestion
- For any questions about the operation of the demos, ask on this repo [with an issue](https://github.com/App-vNext/Polly-Samples/issues/new/choose).
```

> - Any questions about Polly, ask at [Polly](https://www.github.com/App-vNext/Polly).

Reviewer: ```suggestion
- For any questions about Polly, ask in the [Polly repository](https://github.com/App-vNext/Polly/issues/new/choose).
```

> - For full Polly syntax, see Polly repo and the [Polly documentation](https://www.pollydocs.org/).

Reviewer: ```suggestion
- For full Polly syntax, see [Polly repository](https://github.com/App-vNext/Polly#readme) and the [Polly documentation](https://www.pollydocs.org/).
```

| 07 | Wait and retry chaining with circuit breaker by using Pipeline | [Code](PollyDemos/Sync/Demo07_WaitAndRetryNestingCircuitBreakerUsingPipeline.cs) | [Code](PollyDemos/Async/AsyncDemo07_WaitAndRetryNestingCircuitBreakerUsingPipeline.cs) |
| 08 | Fallback, Retry, and CircuitBreaker in a Pipeline | [Code](PollyDemos/Sync/Demo08_Pipeline-Fallback-WaitAndRetry-CircuitBreaker.cs) | [Code](PollyDemos/Async/AsyncDemo08_Pipeline-Fallback-WaitAndRetry-CircuitBreaker.cs) |
| 09 | Fallback, Timeout, and Retry in a Pipeline | [Code](PollyDemos/Sync/Demo09_Pipeline-Fallback-Timeout-WaitAndRetry.cs) | [Code](PollyDemos/Async/AsyncDemo09_Pipeline-Fallback-Timeout-WaitAndRetry.cs) |
| 10 | Without isolation: Faulting calls swamp resources, <br/>also prevent good calls | - | [Code](PollyDemos/Async/AsyncDemo10_SharedConcurrencyLimiter.cs) |
| 11 | With isolation: Faulting calls separated, <br/>do not swamp resources, good calls still succeed | - | [Code](PollyDemos/Async/AsyncDemo11_MultipleConcurrencyLimiters.cs) |

## Want further information?

> - For any questions about the operation of the demos, ask on this repo [with an issue](https://github.com/App-vNext/Polly-Samples/issues/new/choose)).

Reviewer: ```suggestion
- For any questions about the operation of the demos, ask on this repo [with an issue](https://github.com/App-vNext/Polly-Samples/issues/new/choose).
```
